### ISSUE TITLE
Remote SSH: The process tried to write to a nonexistent pipe

### ISSUE THREAD

**Issue by NightJar** [microsoft/vscode-remote-release#977]
Issue Type: <b>Bug</b>

Connect to remote with VSCode insiders and the connection fails "The process tried to write to a nonexistent pipe." This began for me on Friday 12 July 2019 after an restarting VSCode for an update, and previously worked without issue.

At first this seems similar to #525 - but on reading through that thread it appears to be related to path escaping issues - this does not present in the same manner (seems to happen before startup execution reaches that point)

```
Version: 1.37.0-insider (user setup)
Commit: 8fa811108f0f0524c473020ef57b6620f6c201e1
Date: 2019-07-12T09:20:18.712Z
Electron: 4.2.5
Chrome: 69.0.3497.128
Node.js: 10.11.0
V8: 6.9.427.31-electron.0
OS: Windows_NT x64 10.0.17763
```

The "remote" machine is a virtual box instance started via Vagrant, running Arch Linux.

Output:

```
Setting up SSH remote "archdev"
Using commit id "8fa811108f0f0524c473020ef57b6620f6c201e1" and quality "insider" for server
Testing ssh with ssh -V
ssh exited with code: 0
Got stderr from ssh: OpenSSH_for_Windows_7.7p1, LibreSSL 2.6.5
Install and start server if needed
> 
> write: Connection refused
> The process tried to write to a nonexistent pipe.
> 
"install" terminal command done
Received install output: The process tried to write to a nonexistent pipe.
The process tried to write to a nonexistent pipe
```

Interestingly this same remote connection is still functional in the stable release.

Working output from VSCode stable install:

```
Version: 1.36.1 (system setup)
Commit: 2213894ea0415ee8c85c5eea0d0ff81ecc191529
Date: 2019-07-08T22:59:35.033Z
Electron: 4.2.5
Chrome: 69.0.3497.128
Node.js: 10.11.0
V8: 6.9.427.31-electron.0
OS: Windows_NT x64 10.0.17763
```

Output:

```
SSH Resolver called for host: archdev
Setting up SSH remote "archdev"
Using commit id "2213894ea0415ee8c85c5eea0d0ff81ecc191529" and quality "stable" for server
Testing ssh with ssh -V
ssh exited with code: 0
Got stderr from ssh: OpenSSH_for_Windows_7.7p1, LibreSSL 2.6.5
Running script with connection command: "ssh"  -o ClearAllForwardings=true    "archdev" bash
Install and start server if needed
> 
> Running remote connection script
> 
> Found existing installation at /home/vagrant/.vscode-server/bin/2213894ea0415ee8
> c85c5eea0d0ff81ecc191529...
> Found running server...
```

This leads to an _uneducated_ guess that there's an issue with the remote server in 8fa8111 which means it fails to start.

Extension version: 0.44.2
VS Code version: Code - Insiders 1.37.0-insider (8fa811108f0f0524c473020ef57b6620f6c201e1, 2019-07-12T09:20:18.712Z)
OS version: Windows_NT x64 10.0.17763

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-4712HQ CPU @ 2.30GHz (8 x 2295)|
|GPU Status|2d_canvas: enabled<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>native_gpu_memory_buffers: disabled_software<br>oop_rasterization: disabled_off<br>protected_video_decode: unavailable_off<br>rasterization: enabled<br>skia_deferred_display_list: disabled_off<br>skia_renderer: disabled_off<br>surface_synchronization: enabled_on<br>video_decode: enabled<br>viz_display_compositor: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|15.91GB (8.14GB free)|
|Process Argv||
|Screen Reader|no|
|VM|0%|
</details>

**Comment by roblourens** [comment]
What version of the extension did you use in Insiders? Was it updated to 0.44.2? It looks like you trimmed off that line of the log.

**Comment by NightJar** [comment]
Ooh, sorry. I had assumed that was the hash that is present, re-reading it now it looks like the version hash of VSCode itself.

The issue was created (drafted) from within the insiders environment that was presenting this issue... although I had a little trouble getting it to report an issue with an extension at first, as it found none installed. But that went away when I told the Remote extension to disconnect (from it's unconnected session).

The stamp at the base of the issue report should be correct to the working environment manifesting the issue :)

I can double check this when I'm next working at that machine.

**Comment by NightJar** [comment]
🤔 Well, ok. Today is a new story!

```
remote-ssh@0.44.2
win32 x64
SSH Resolver called for "ssh-remote+archdev", attempt 1
SSH Resolver called for host: archdev
Setting up SSH remote "archdev"
Using commit id "8fa811108f0f0524c473020ef57b6620f6c201e1" and quality "insider" for server
Testing ssh with ssh -V
ssh exited with code: 0
Got stderr from ssh: OpenSSH_for_Windows_7.7p1, LibreSSL 2.6.5
Running script with connection command: "ssh"  -o ClearAllForwardings=true    "archdev" bash
Install and start server if needed
> 
> Running remote connection script
> 
> Found old VS Code install 41691474e5fa22a279d49fc4b58af66bc11ce41b, attempting t
> o clean up
> 
> Deleting old install from /home/vagrant/.vscode-server-insiders/bin/41691474e5fa
> 22a279d49fc4b58af66bc11ce41b
> 
> Installing to /home/vagrant/.vscode-server-insiders/bin/8fa811108f0f0524c473020e
> f57b6620f6c201e1...
> Downloading with wget
> 
> Download complete
> 
> SHELL=/bin/bash
> PWD=/home/vagrant
> LOGNAME=vagrant
> XDG_SESSION_TYPE=tty
> HOME=/home/vagrant
> VSCODE_AGENT_FOLDER=/home/vagrant/.vscode-server-insiders
> SSH_CONNECTION=10.0.2.2 58206 10.0.2.15 22
> XDG_SESSION_CLASS=user
> USER=vagrant
> SHLVL=1
> XDG_SESSION_ID=6
> XDG_RUNTIME_DIR=/run/user/1000
> SSH_CLIENT=10.0.2.2 58206 22
> PATH=/usr/local/sbin:/usr/local/bin:/usr/bin
> DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/bus
> MAIL=/var/spool/mail/vagrant
> _=/usr/bin/printenv
> OLDPWD=/home/vagrant/.vscode-server-insiders/bin/8fa811108f0f0524c473020ef57b662
> 0f6c201e1
> Starting server...
> Waiting for server log...
> 
> Waiting for server log...
> 
> 
> *
> * Reminder: You may only use this software with Visual Studio family products,
> * as described in the license (https://go.microsoft.com/fwlink/?linkid=2077057)
> *
> 
> 
> 3e0a0cf7-21ed-4e8e-8ac0-848ee034109c==36405==44283==
> 
"install" terminal command done
Received install output: 3e0a0cf7-21ed-4e8e-8ac0-848ee034109c==36405==44283==
Server is listening on port 36405
Running script with connection command: "ssh" -L 127.0.0.1:36405:127.0.0.1:36405     "archdev" bash
Spawning tunnel with: type "C:\Users\DWAGST~1\AppData\Local\Temp\vscode-linux-multi-line-command-archdev-578389196.sh" | "ssh" -L 127.0.0.1:36405:127.0.0.1:36405     "archdev" bash
> 
> Connected to SSH Host - Please do not close this terminal
> 
Spawned SSH tunnel between local port 36405 and remote port 36405
Waiting for ssh tunnel to be ready
Tunneling remote port 36405 to local port 36405
Resolved "ssh-remote+archdev" to "127.0.0.1:36405"
------
```

Something new happened, in that where the process used to fail after trying to write to a non-existent pipe, it now detected an old version, removed it, updated itself, and connected successfully 🎉  \o/

Perhaps it comes down to a version mis-match between what already existed on the server, and the recently updated local copy that was attempting to connect to it. Unsure, either way it has rectified itself with an update.

I see my insiders copy is also waiting on an update, so I'll see if everything runs smooth after that too...

**Comment by NightJar** [comment]
Yes, another update, and still running successfully.
I note it repeated the self update steps again, which were notably absent when I opened this issue :)

```
> Deleting old install from /home/vagrant/.vscode-server-insiders/bin/a122c88b4e56
> 318f778d9d2e3800395ee18ecba8
> Installing to /home/vagrant/.vscode-server-insiders/bin/e1d3dd53d17fb1529a002e4d
> 6fb066db0a0bd385...
```

Thanks for taking the time to respond to my issue :)

**Comment by jelliedpizza** [comment]
Hey I got the same error today. So how did you solve the "The process tried to write to a nonexistent pipe" error?

**Comment by NightJar** [comment]
Hi @jelliedpizza I also got this again today... however the reason was quite obvious (this time): My virtual box was not running :)
To that end I recommend testing that the machine is reachable via a normal SSH connection (e.g. if it is not a virtual machine).

That aside, I solve the issue with an update. Please note this issue was specific to the `insiders` build, which updates frequently.

**Comment by sagerenn** [comment]
I have the problem too. I connect successfully to the host with one computer, but failed to open it with another computer.

```
> Password: 
"install" wrote data to terminal: "**********"
> 
> 
> WSASendCB - ERROR: broken assumption, io:0000022A1D594680, sent:84, remaining:0
> 
> The process tried to write to a nonexistent pipe.
> 
"install" terminal command done
Received install output: The process tried to write to a nonexistent pipe.
The process tried to write to a nonexistent pipe
```

**Comment by NightJar** [comment]
It seems from that snippet that the connection has failed to establish (e.g. incorrect password) rather than some other kind of error.

You mention that there are two client machines in play - check that both are using the same version of VSCode and remote-ssh plugin. This is critical as you mention that one works, and the other doesn't - this hints that the problem probably lies outside of VSCode and the remote-ssh plugin e.g. configuration of either the client or the host.

On the host side; the snippet you posted also reports password authentication. I recommend you set up public key authorisation and test again. The configuration will need to be the same on both machines for an accurate test, and will also be good as steps to reproduce the issue.

If everything is verified to be the same between client environments, and nothing on the host is actively blocking connection (e.g. the remote server has password authentication disabled for SSH) then I would recommend opening a new issue, taking care to list accurate steps to reproduce the problem.

**Comment by sagerenn** [comment]
Maybe the computer that doesn't work has some limitations. It works when I try it again with another virtual machine.
